File: bootcamp-c-language/Quest04/ex05/my_spaceship.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h> // for malloc

char* my_spaceship(char* flight_path) {
  int x = 0;
  int y = 0;
  char* direction = "up";

  // Iterate through the string and update the x, y coordinates and direction
  // based on the letters in the flight path
  for (int i = 0; i < strlen(flight_path); i++) {
    if (flight_path[i] == 'R') {
      if (strcmp(direction, "up") == 0) {
        direction = "right";
      }
      else if (strcmp(direction, "right") == 0) {
        direction = "down";
      }
      else if (strcmp(direction, "down") == 0) {
        direction = "left";
      }
      else if (strcmp(direction, "left") == 0) {
        direction = "up";
      }
    }
    else if (flight_path[i] == 'L') {
      if (strcmp(direction, "up") == 0) {
        direction = "left";
      }
      else if (strcmp(direction, "left") == 0) {
        direction = "down";
      }
      else if (strcmp(direction, "down") == 0) {
        direction = "right";
      }
      else if (strcmp(direction, "right") == 0) {
        direction = "up";
      }
    }
    else if (flight_path[i] == 'A') {
      if (strcmp(direction, "up") == 0) {
        y--;
      }
      else if (strcmp(direction, "right") == 0) {
        x++;
      }
      else if (strcmp(direction, "down") == 0) {
        y++;
      }
      else if (strcmp(direction, "left") == 0) {
        x--;
      }
    }
  }

  // Allocate memory for the result string
  // (we need to allocate enough space to hold the largest possible result string)
  char* result = malloc(64 * sizeof(char));

  // Create the string to return in the required format
  sprintf(result, "{x: %d, y: %d, direction: '%s'}", x, y, direction);
  return result;
}

//  int main() {
//   // Test the my_spaceship function
//   char* flight_path = "AAAA";
//   char* result = my_spaceship(flight_path);
//   printf("%s", result);

//   return 0;
// }

File: bootcamp-c-language/my_spaceship/ex00/my_spaceship.js
function turn_right(spaceship){
  if(spaceship["direction"] == "up"){
      spaceship["direction"] = "right";
  } else if (spaceship["direction"] = "left"){
      spaceship["direction"] = "up";
  } else if (spaceship["direction"] = "down"){
      spaceship["direction"] = "left";
  } else if (spaceship["direction"] = "right"){
      spaceship["direction"] = "down";
  }
}

function turn_left(spaceship){
  if(spaceship["direction"] == "up"){
      spaceship["direction"] = "left";
  } else if (spaceship["direction"] = "left"){
      spaceship["direction"] = "down";
  } else if (spaceship["direction"] = "down"){
      spaceship["direction"] = "right";
  } else if (spaceship["direction"] = "right"){
      spaceship["direction"] = "up";
  }
}

function move(spaceship){
  if(spaceship["direction"] == "up"){
      spaceship["y"] -= 1;
  } else if (spaceship["direction"] = "left"){
      spaceship["x"] -= 1;
  } else if (spaceship["direction"] = "down"){
      spaceship["y"] += 1;
  } else if (spaceship["direction"] = "right"){
      spaceship["x"] += 1;
  }
}

function my_spaceship(params){
  let index = 0;
  let spaceship = {x: 0, y: 0, direction: "up"}
  while(index < params.length){
      let letter = params[index];
      if (letter == "A") {
          move(spaceship);
      }
      else if (letter == "R"){
          turn_right(spaceship);
      }
      else if (letter == "L"){
          turn_left(spaceship);
      }
      index += 1;
  }
  return spaceship;
}

console.log(my_spaceship("RAALALL"))
console.log(my_spaceship("AAAA"))
console.log(my_spaceship(""))
console.log(my_spaceship("RAARA"))